libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp

```diff
@@ -0,0 +1,286 @@
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    float sum_cm = 0;
+    uint16_t count = 0;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+        if (b < 0) {
+            continue;
+        }
+        if (parse_byte(b)) {
+            count++;
+            sum_cm += distance_cm;
+        }
+    }
+
+    if (count > 0) {
+        // return average distance of readings
+        reading_cm = sum_cm / count;
+
+        // request another distance
+        send_message(BLDPIN_MSGID_DISTANCE_SIMPLE, nullptr, 0);
+
+        return true;
+    }
+
+    // no readings so return false
+    return false;
+}
+
+// process one byte received on serial port
+// returns true if a distance message has been successfully parsed
+// state is stored in msg structure
+bool AP_RangeFinder_BLPing::parse_byte(uint8_t b)
+{
+    bool got_distance = false;
+
+    // process byte depending upon current state
+    switch (msg.state) {
+
+    case MSG_STATE_HEADER1:
+        if (b == BLPING_FRAME_HEADER1) {
+            msg.crc_expected = BLPING_FRAME_HEADER1;
+            msg.state = MSG_STATE_HEADER2;
+        }
+        break;
+
+    case MSG_STATE_HEADER2:
+        if (b == BLPING_FRAME_HEADER2) {
+            msg.crc_expected += BLPING_FRAME_HEADER2;
+            msg.state = MSG_STATE_LEN_L;
+        } else {
+            msg.state = MSG_STATE_HEADER1;
+        }
+        break;
+
+    case MSG_STATE_LEN_L:
+        msg.payload_len = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_LEN_H;
+        break;
+
+    case MSG_STATE_LEN_H:
+        msg.payload_len += ((uint16_t)b << 8);
+        msg.payload_recv = 0;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_MSG_ID_L;
+        break;
+
+    case MSG_STATE_MSG_ID_L:
+        msg.msgid = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_MSG_ID_H;
+        break;
+
+    case MSG_STATE_MSG_ID_H:
+        msg.msgid += ((uint16_t)b << 8);
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_SRC_ID;
+        break;
+
+    case MSG_STATE_SRC_ID:
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_DST_ID;
+        break;
+
+    case MSG_STATE_DST_ID:
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_PAYLOAD;
+        break;
+
+    case MSG_STATE_PAYLOAD:
+        if (msg.payload_recv < msg.payload_len) {
+            msg.payload[msg.payload_recv] = b;
+            msg.payload_recv++;
+            msg.crc_expected += b;
+        }
+        if (msg.payload_recv == msg.payload_len) {
+            msg.state = MSG_STATE_CRC_L;
+        }
+        break;
+
+    case MSG_STATE_CRC_L:
+        msg.crc = b;
+        msg.state = MSG_STATE_CRC_H;
+        break;
+
+    case MSG_STATE_CRC_H:
+        msg.crc += ((uint16_t)b << 8);
+        msg.state = MSG_STATE_HEADER1;
+        if (msg.crc_expected == msg.crc) {
```

Comment:
No diagnostics about parse errors in this driver.  *shrug*

Reply:
yes, no recording of parsing errors.. at the risk of passing the buck, I don't think we have reporting of parsing errors from other lidar either..